lib/internal/modules/esm/assets/load.js

```diff
@@ -0,0 +1,241 @@
+  HTTPSGet = require('https').get;
+  return HTTPSGet(url, opts);
+}
+
+function HTTPGet(url, opts) {
+  // eslint-disable-next-line no-func-assign
+  HTTPGet = require('http').get;
+  return HTTPGet(url, opts);
+}
+
+function dnsLookup(name, opts) {
+  // eslint-disable-next-line no-func-assign
+  dnsLookup = require('dns/promises').lookup;
+  return dnsLookup(name, opts);
+}
+
+/**
+ * @param {URL} parsed
+ * @returns {Promise<CacheEntry> | CacheEntry}
+ */
+function fetchWithRedirects(parsed) {
+  const existing = getCache.get(parsed.href);
+  if (existing) {
+    return existing;
+  }
+  const handler = parsed.protocol === 'http:' ? HTTPGet : HTTPSGet;
+  const result = new Promise((fulfill, reject) => {
+    const req = handler(parsed, {
+      headers: {
+        Accept: '*/*'
+      }
+    })
+    .on('error', reject)
+    .on('response', (res) => {
+      function dispose() {
+        req.destroy();
+        res.destroy();
+      }
+      if (res.statusCode >= 300 && res.statusCode <= 303) {
+        if (res.headers.location) {
+          dispose();
+          const location = new URL(res.headers.location, parsed);
+          if (location.protocol !== 'http:' && location.protocol !== 'https:') {
+            reject(new ERR_UNSUPPORTED_ESM_NETWORK_IMPORT_SPECIFIER(
+              res.headers.location,
+              parsed.href,
+              'cannot redirect to non-network location'));
+            return;
+          }
+          return PromiseResolve(
+            fetchWithRedirects(location)
+          ).then((entry) => {
+            getCache.set(parsed.href, entry);
+            fulfill(entry);
+          });
+
+        }
+      }
+      if (res.statusCode > 303 && res.statusCode < 200) {
+        dispose();
+        reject(
+          new ERR_UNSUPPORTED_ESM_NETWORK_BAD_RESPONSE(
+            'HTTP response returned status code of ' + res.statusCode));
+        return;
+      }
+      const { headers } = res;
+      const contentType = headers['content-type'];
+      if (!contentType) {
+        dispose();
+        reject(new ERR_UNSUPPORTED_ESM_NETWORK_BAD_RESPONSE(
+          'content-type is required'));
+        return;
+      }
+      /**
+       * @type {CacheEntry}
+       */
+      const entry = {
+        resolvedHREF: parsed.href,
+        headers: {
+          'content-type': res.headers['content-type']
+        },
+        body: new Promise((f, r) => {
+          res.on('error', (e) => {
+            dispose();
+            reject(e);
+            r(e);
+          });
+          const buffers = [];
+          let size = 0;
+          res.on('data', (d) => {
+            ArrayPrototypePush(buffers, d);
+            size += d.length;
+          });
+          res.on('end', () => {
+            const body = entry.body = /** @type {Buffer} */(
+              BufferConcat(buffers, size)
+            );
+            f(body);
+          });
+        }),
+      };
+      getCache.set(parsed.href, entry);
```

Comment:
It seems we might be keeping the http response within the cache. I would not recommend to cache the http objects forever.

Reply:
we should keep them around for some period of time, and given that the offline cache would persist them forever; what would be the recommendation